source/common/http/header_map_impl.cc

```diff
@@ -21,190 +21,103 @@ constexpr size_t MinDynamicCapacity{32};
 // This includes the NULL (StringUtil::itoa technically only needs 21).
 constexpr size_t MaxIntegerLength{32};
 
-uint64_t newCapacity(uint32_t existing_capacity, uint32_t size_to_append) {
-  return (static_cast<uint64_t>(existing_capacity) + size_to_append) * 2;
-}
-
 void validateCapacity(uint64_t new_capacity) {
   // If the resizing will cause buffer overflow due to hitting uint32_t::max, an OOM is likely
   // imminent. Fast-fail rather than allow a buffer overflow attack (issue #1421)
   RELEASE_ASSERT(new_capacity <= std::numeric_limits<uint32_t>::max(),
                  "Trying to allocate overly large headers.");
   ASSERT(new_capacity >= MinDynamicCapacity);
 }
-
 } // namespace
 
-HeaderString::HeaderString() : type_(Type::Inline) {
-  buffer_.dynamic_ = inline_buffer_;
-  clear();
-  static_assert(sizeof(inline_buffer_) >= MaxIntegerLength, "");
+// Initialize as a Type::Inline
+HeaderString::HeaderString() : buffer_(absl::InlinedVector<char, 128>()), string_length_(0) {
+  static_assert(sizeof(buffer_) >= MaxIntegerLength, "");
   static_assert(MinDynamicCapacity >= MaxIntegerLength, "");
   ASSERT(valid());
 }
 
-HeaderString::HeaderString(const LowerCaseString& ref_value) : type_(Type::Reference) {
-  buffer_.ref_ = ref_value.get().c_str();
-  string_length_ = ref_value.get().size();
+// Initialize as a Type::Reference
+HeaderString::HeaderString(const LowerCaseString& ref_value)
+    : buffer_(absl::string_view(ref_value.get().c_str(), ref_value.get().size())),
+      string_length_(ref_value.get().size()) {
   ASSERT(valid());
 }
 
-HeaderString::HeaderString(absl::string_view ref_value) : type_(Type::Reference) {
-  buffer_.ref_ = ref_value.data();
-  string_length_ = ref_value.size();
+// Initialize as a Type::Reference
+HeaderString::HeaderString(absl::string_view ref_value)
+    : buffer_(ref_value), string_length_(ref_value.size()) {
   ASSERT(valid());
 }
 
-HeaderString::HeaderString(HeaderString&& move_value) noexcept {
-  type_ = move_value.type_;
-  string_length_ = move_value.string_length_;
-  switch (move_value.type_) {
-  case Type::Reference: {
-    buffer_.ref_ = move_value.buffer_.ref_;
-    break;
-  }
-  case Type::Dynamic: {
-    // When we move a dynamic header, we switch the moved header back to its default state (inline).
-    buffer_.dynamic_ = move_value.buffer_.dynamic_;
-    dynamic_capacity_ = move_value.dynamic_capacity_;
-    move_value.type_ = Type::Inline;
-    move_value.buffer_.dynamic_ = move_value.inline_buffer_;
-    move_value.clear();
-    break;
-  }
-  case Type::Inline: {
-    buffer_.dynamic_ = inline_buffer_;
-    memcpy(inline_buffer_, move_value.inline_buffer_, string_length_);
-    move_value.string_length_ = 0;
-    break;
-  }
-  }
+HeaderString::HeaderString(HeaderString&& move_value) noexcept
+    : buffer_(move_value.buffer_), string_length_(move_value.string_length_) {
```

Comment:
I think this might copy. I would store the right hand side string length in a local variable and then do an explicit std::move?

Reply:
Fixed.

Reply:
Also removed `string_length_`, as it is not really necessary anymore.